[Cases] Notify assignees when assigned to a case

cnasikas opened the pull request:
## Summary

The PR adds the ability to notify users by email when assigned to a case. A user is:
- Not notified if he/she assigns themselves
- Notified if added as an assignee to a case
- Not notified if removed from a case

I did not add integration tests due to the complexity of simulating an email server. I added unit test coverage. If integration test coverage is needed we can add the tests on another PR.

Depends on: https://github.com/elastic/kibana/pull/143303
Fixes: https://github.com/elastic/kibana/issues/142307

## Email screenshot

<img width="365" alt="Screenshot 2022-11-08 at 7 05 39 PM" src="https://user-images.githubusercontent.com/7871006/200629737-86e56092-bff7-4fee-954a-d925c8b0c3d3.png">


@shanisagiv1 @lcawl What do you think about the content of the email (see screenshot)?

## Testing

1. Put the following in your `kibana.yml`:

```
notifications.connectors.default.email: 'mail-dev'

xpack.actions.preconfigured:
  mail-dev:
    name: preconfigured-email-notification-maildev
    actionTypeId: .email
    config:
      service: other
      from: mlr-test-sink@elastic.co
      host: localhost
      port: 1025
      secure: false
      hasAuth: false
```

2. Install [`maildev`](https://www.npmjs.com/package/maildev): `npm install -g maildev`
3. Run `maildev`: `maildev`
4. Open MailDev's web interface (http://0.0.0.0:1080/)
5. Create a case and assign users. You should see the emails in the MailDev inbox. 
6. Update the assignees of a case. You should see the emails in the MailDev inbox. 

Note: If you assign yourself you should not see an email. If you delete an assignee you should not see an email.

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

## Release notes

Notify users by email when assigned to a case

elasticmachine commented:
Pinging @elastic/response-ops-cases (Feature:Cases)

shanisagiv1 commented:
a few comments:
- can we call it "Elastic Kibana Case" instead of "Elastic Case"
- a question for @lcawl  - "you got assigned" or "you have been assigned to" - what is more accurate?
- "view case" URL --> "View the case details"

not related to the content - how the FROM address is defined?

cnasikas commented:
>can we call it "Elastic Kibana Case" instead of "Elastic Case"
Ofc!

>a question for @lcawl - "you got assigned" or "you have been assigned to" - what is more accurate?

This is my mistake. @lcawl suggested, "You are assigned to a case". I forgot to update.

>"view case" URL --> "View the case details"
Ok!

>not related to the content - how the FROM address is defined?

From the email connector configuration. The user decides what the FROM address will be.

cnasikas commented:
@shanisagiv1, @gsoldevila found out that in the cloud the `from` is set automatically to [`noreply@alerts.elastic.co`](https://github.com/elastic/cloud/blob/85406616045911ebe5d3476e35c409c8fc88c8f7/scala-services/adminconsole/src/main/scala/no/found/adminconsole/settings/DeploymentSmtpSettings.scala#L33). On-prem the users can configure it as they want.

lcawl commented on x-pack/plugins/cases/server/services/notifications/email_notification_service.test.ts:
```diff
@@ -0,0 +1,197 @@
+        ],
+      },
+      message:
+        'You got assigned to an Elastic Case.\r\n\r\nTitle: Super Bad Security Issue\r\n\r\nStatus: open\r\n\r\nSeverity: low\r\n\r\nTags: defacement\r\n\r\n\r\n\r\n[View case](https://example.com/app/security/cases/mock-id-1)',
```
```suggestion
        'You are assigned to an Elastic Case.\r\n\r\nTitle: Super Bad Security Issue\r\n\r\nStatus: open\r\n\r\nSeverity: low\r\n\r\nTags: defacement\r\n\r\n\r\n\r\n[View case](https://example.com/app/security/cases/mock-id-1)',
```

cnasikas replied:
Thanks Lisa, I forgot to do the change!

lcawl commented on x-pack/plugins/cases/server/services/notifications/email_notification_service.test.ts:
```diff
@@ -0,0 +1,197 @@
+        ],
+      },
+      message:
+        'You got assigned to an Elastic Case.\r\n\r\nTitle: Super Bad Security Issue\r\n\r\nStatus: open\r\n\r\nSeverity: low\r\n\r\nTags: defacement\r\n\r\n\r\n\r\n[View case](https://example.com/app/security/cases/mock-id-1)',
```
```suggestion
        'You are assigned to an Elastic Case.\r\n\r\nTitle: Super Bad Security Issue\r\n\r\nStatus: open\r\n\r\nSeverity: low\r\n\r\nTags: defacement\r\n\r\n\r\n\r\n[View case](https://example.com/app/security/cases/mock-id-1)',
```

lcawl commented on x-pack/plugins/cases/server/services/notifications/email_notification_service.test.ts:
```diff
@@ -0,0 +1,197 @@
+        ],
+      },
+      message:
+        'You got assigned to an Elastic Case.\r\n\r\nTitle: Super Bad Security Issue\r\n\r\nStatus: open\r\n\r\nSeverity: low\r\n\r\nTags: defacement\r\n\r\n\r\n\r\n[View case](https://example.com/app/security/cases/mock-id-1)',
```
```suggestion
        'You are assigned to an Elastic Case.\r\n\r\nTitle: Super Bad Security Issue\r\n\r\nStatus: open\r\n\r\nSeverity: low\r\n\r\nTags: defacement\r\n\r\n\r\n\r\n[View case](https://example.com/app/security/cases/mock-id-1)',
```

lcawl commented on x-pack/plugins/cases/server/services/notifications/email_notification_service.test.ts:
```diff
@@ -0,0 +1,197 @@
+        ],
+      },
+      message:
+        'You got assigned to an Elastic Case.\r\n\r\nTitle: Super Bad Security Issue\r\n\r\nStatus: open\r\n\r\nSeverity: low\r\n\r\nTags: defacement\r\n\r\n\r\n\r\n[View case](https://example.com/app/security/cases/mock-id-1)',
```
```suggestion
        'You are assigned to an Elastic Case.\r\n\r\nTitle: Super Bad Security Issue\r\n\r\nStatus: open\r\n\r\nSeverity: low\r\n\r\nTags: defacement\r\n\r\n\r\n\r\n[View case](https://example.com/app/security/cases/mock-id-1)',
```

lcawl commented on x-pack/plugins/cases/server/services/notifications/email_notification_service.test.ts:
```diff
@@ -0,0 +1,197 @@
+        ],
+      },
+      message:
+        'You got assigned to an Elastic Case.\r\n\r\nTitle: Super Bad Security Issue\r\n\r\nStatus: open\r\n\r\nSeverity: low\r\n\r\nTags: defacement\r\n\r\n',
```
```suggestion
        'You are assigned to an Elastic Case.\r\n\r\nTitle: Super Bad Security Issue\r\n\r\nStatus: open\r\n\r\nSeverity: low\r\n\r\nTags: defacement\r\n\r\n',
```

lcawl commented on x-pack/plugins/cases/server/services/notifications/email_notification_service.ts:
```diff
@@ -0,0 +1,118 @@
+
+  private getMessage(theCase: CaseSavedObject) {
+    const lineBreak = '\r\n\r\n';
+    let message = `You got assigned to an Elastic Case.${lineBreak}`;
```
```suggestion
    let message = `You are assigned to an Elastic Case.${lineBreak}`;
```

lcawl reviewed (approved):
Suggest replacing "got" in the message, otherwise text LGTM

jonathan-buttner commented on x-pack/plugins/cases/server/client/cases/update.ts:
```diff
@@ -437,6 +452,16 @@ export const update = async (
+      user
+    );
+
+    if (casesAndAssigneesToNotifyForAssignment.length > 0) {
```
How about we move the if check inside of the `bulkNotifyAssignees` and just have it return if the passed in value is an empty array?

jonathan-buttner commented on x-pack/plugins/cases/server/client/cases/update.ts:
```diff
@@ -497,3 +522,48 @@ const patchCases = async ({
+    }
+
+    return acc;
+  }, [] as Array<{ assignees: CaseAssignees; theCase: CaseSavedObject }>);
```
nit: I think we can avoid the cast if we add this definition to the `reduce<Array<{ assignees: CaseAssignees; theCase: CaseSavedObject }>>`

jonathan-buttner commented on x-pack/plugins/cases/server/services/notifications/email_notification_service.test.ts:
```diff
@@ -0,0 +1,197 @@
+  const caseSO = mockCases[0];
+  const assignees = userProfiles.map((userProfile) => ({ uid: userProfile.uid }));
+
+  const notifications = {
```
Just curious, does the notification plugin return a mock that we could leverage here?

cnasikas replied:
Good catch! When I wrote the code they did not but they added it later. I will update to include it.

jonathan-buttner commented on x-pack/plugins/cases/server/services/notifications/email_notification_service.test.ts:
```diff
@@ -0,0 +1,197 @@
+    ]);
+
+    await emailNotificationService.notifyAssignees({
+      assignees: [...assignees, { uid: assignees[0].uid }],
```
nit: to make it clearer maybe we should remove the duplicate assignee?

jonathan-buttner commented on x-pack/plugins/cases/server/services/notifications/email_notification_service.ts:
```diff
@@ -0,0 +1,118 @@
+    this.publicBaseUrl = publicBaseUrl;
+  }
+
+  private getTitle(theCase: CaseSavedObject) {
```
nit: we could make this a `private static getTitle` method.

jonathan-buttner commented on x-pack/plugins/cases/server/services/notifications/email_notification_service.ts:
```diff
@@ -0,0 +1,118 @@
+  }
+
+  public async notifyAssignees({ assignees, theCase }: NotifyArgs) {
+    if (!this.notifications.isEmailServiceAvailable()) {
```
Don't know if it's possible for `isEmailServiceAvailable()` to throw an error but maybe we should move it into the `try`?

jonathan-buttner commented on x-pack/plugins/cases/server/services/notifications/email_notification_service.ts:
```diff
@@ -0,0 +1,118 @@
+
+  public async bulkNotifyAssignees(args: NotifyArgs[]) {
+    await Promise.all(
+      args.map(({ assignees, theCase }) => this.notifyAssignees({ assignees, theCase }))
```
Let's use the `pMap` to ensure we don't make to many concurrent requests.

jonathan-buttner reviewed (commented):
Code looks great, left a few comments. I still need to test it. As for integration tests, it looks like the email connector uses a mock email server to receive its emails. Do you think we could implement the same thing? We can do it as a follow up PR.

cnasikas commented:
> Code looks great, left a few comments. I still need to test it. As for integration tests, it looks like the email connector uses a mock email server to receive its emails. Do you think we could implement the same thing? We can do it as a follow up PR.

Thank you @jonathan-buttner for the feedback. I addressed all the suggestions. The email server is not mocked.  The `service: "__json"` is passed as a param when the connector is executed. That is a special "service" that does not actually send the email but it uses nodemailer's jsonTransport to dry run sending emails. In our case, it is not possible to get the output of the executor as it is run in the task manager and the response is not written to any log. We talked about it with @pmuellr and to do it we need to use [`maildev`](https://www.npmjs.com/package/maildev) to mock the email server. Then we need to check for new emails every 1s and save them in memory to be able to later search them when needed from a test. Given the limited time and the amount of work is need to simulate the email server I would prefer to do it on another PR if you do not mind.

jonathan-buttner reviewed (approved):
Great work! Tested and looks good.

EricDavisX commented:
Wanted to share that I have done a quick test and seen the 'Cases Assignment will send an email notification' feature working via Cloud-QA deploy of latest 8.6 snapshot code, with the current known requirement that the users that shall receive email notifications are whitelisted in cloud, per [this meaning](https://www.elastic.co/guide/en/cloud/current/ec-watcher.html#ec-watcher-allowlist)
